bundles/io/org.eclipse.smarthome.io.security/src/main/java/org/eclipse/smarthome/io/security/api/SecretsVaultService.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.security.api;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The {@link SecretsVaultService} can be used to store and retrieve secrets.
+ *
+ * Secrets can only be retrieved by the bundle that stored them and by default
+ * will be encrypted and and stored on disk immediately.
+ *
+ * Because of the way how the String class is handled internally, you are strongly
+ * encouraged to use char or byte arrays or the ByteBuffer class
+ * for keeping encrypted secrets in memory.
+ *
+ * The encryption algorithm, the storage location and format depend on the implementation.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public interface SecretsVaultService {
+    /**
+     * Retrieve a secret, identified by the given key.
+     *
+     * @param key An identifier
+     * @return Either the secret as ByteBuffer that was stored by this bundle under the given key
+     *         or null if no secret could be found.
+     * @throws IOException If the underlying storage file can't be opened or read.
+     * @throws GeneralSecurityException If the decryption algorithm fails, this exception is thrown
+     */
+    @Nullable
+    ByteBuffer retrieve(String key) throws IOException, GeneralSecurityException;
```

Comment:
I still do not understand how this retrieving of  a secret works. As a user of this new security API if I want to retrieve a secret for my bundle what do I have to do?

1. List<String> mySecrets = SecretsVaultManagement.getAllKeys("myBundle")
2. for(String secret : mySecrets) { SecretsVaultService.retrieve(secret) }

If this is correct for `retrieve`, how does `add` work? Do I use `SecretsVaultManagement.modify` also to add new keys? If so we should rename it to something like `addOrModify` to make it more transparent.

Reply:
As a consumer of the api, you know your keys. You probably would not iterate over all your stored secrets.

Add only adds and should either return with a Boolean or throw if the key is already existing.

If the consumer needs to modify, it would delete and add again.

---

The rest API can't add with the current proposal, I forgot about that. We probably rename modify to "set" and allow to add a secret as well with an empty old secret if there's nothing saved so far under the current key.

Reply:
> If the consumer needs to modify, it would delete and add again.

Why do we offer `SecretsVaultManagement.modify` (or `set` after the change below) then?

> The rest API can't add with the current proposal, I forgot about that. We probably rename modify to "set" and allow to add a secret as well with an empty old secret if there's nothing saved so far under the current key.

So you suggest to rename `SecretsVaultManagement.modify` to `SecretsVaultManagement.set` and allow for adding new secrets and overwriting existing ones (with matching oldSecret)? That's what I meant, so fine with me.

Reply:
To answer your question: The service interface is to be used within a bundle for that bundles secrets. The old secret is not necessary to modify a secret here. 

The management interface can be used by any bundle but might be security restricted internally and the bundle name needs to be supplied as well as old secret to change a secret.

Reply:
I understand your separation but how do we ensure that the implementation of the service interface is not called by someone else? I mean this implementation has a `delete` and `add` method that are public, right? But maybe this is a stupid question because we would have to expose these methods from the implementation somehow first before others can use it?

Reply:
I thought about the OSGi service factory. If a bundle requests the secret vault service, an specific instance for that bundle is created. I haven't tested this with code yet.

Reply:
Could you please give it a try and give some details on how that could work? If it doesn't I would clearly vote for an additional argument to the `delete` method which is the `oldValue` and only delete it if it matches.